File: codes/lua/custom_error_and_error_handling.lua
print("\n-- Custom Error and Error Handling in Lua")

function give_me_respect(argument)
    -- Custom Error
    if (type(argument) ~= "string") then error("Argument should be a string") end
    if (argument ~= "respect") then error('You should give me "respect"!') end
    return 'Thank you for giving me "respect"!'
end

-- Error Handling
is_success, any_result = pcall(give_me_respect, "boo!")
if (is_success == true) then
    print(any_result)
else
    print(any_result)
end
print("I'm sorry!")

-- Without Error Handling
response = give_me_respect("boo!")
-- this will never executed
print(response)
print("I'm sorry!")

File: codes/perl/custom_error_and_error_handling.pl
use strict;
use warnings;
use Try::Tiny; # https://metacpan.org/pod/Try::Tiny

print("\n# Custom Error and Error Handling in Perl", "\n");

sub give_me_respect {
    my ($argument) = @_;
    # Custom Error
    die("Error: You should give me \"respect\"!") if ($argument ne "respect");
    return "Thank you for giving me \"respect\"!";
}

my $response;

# Error Handling using 3rd party module "Try::Tiny"
# https://metacpan.org/pod/Try::Tiny
try {
    $response = give_me_respect("boo!");
    print($response, "\n");
} catch {
    print($_, "\n");
};
print("I'm sorry!", "\n");

# Without Error Handling
$response = give_me_respect("boo!");
# this will never executed
print($response, "\n");
print("I'm sorry!");
